sweep/tx_input_set.go

```diff
@@ -0,0 +1,124 @@
+	feePerKW chainfee.SatPerKWeight
+
+	// inputs is the set of tx inputs.
+	inputs []input.Input
+
+	// dustLimit is the minimum output value of the tx.
+	dustLimit btcutil.Amount
+}
+
+// newTxInputSet constructs a new, empty input set.
+func newTxInputSet(feePerKW, relayFee chainfee.SatPerKWeight) *txInputSet {
+	dustLimit := txrules.GetDustThreshold(
+		input.P2WPKHSize,
+		btcutil.Amount(relayFee.FeePerKVByte()),
+	)
+
+	b := txInputSet{
+		feePerKW:  feePerKW,
+		dustLimit: dustLimit,
+	}
+
+	// Add the sweep tx output to the weight estimate.
+	b.weightEstimate.AddP2WKHOutput()
+
+	return &b
+}
+
+// dustLimitReached returns true if we've accumulated enough inputs to meet the
+// dust limit.
+func (b *txInputSet) dustLimitReached() bool {
+	return b.outputValue >= b.dustLimit
+}
+
+// add adds a new input to the set. It returns a bool indicating whether the
+// input was added to the set. An input is rejected if it decreases the tx
+// output value after paying fees.
+func (b *txInputSet) add(input input.Input) bool {
+	// Can ignore error, because it has already been checked when
+	// calculating the yields.
+	size, isNestedP2SH, _ := input.WitnessType().SizeUpperBound()
+
+	// Keep a running weight estimate of the input set.
+	if isNestedP2SH {
+		b.weightEstimate.AddNestedP2WSHInput(size)
+	} else {
+		b.weightEstimate.AddWitnessInput(size)
+	}
+
+	value := btcutil.Amount(input.SignDesc().Output.Value)
+	newInputTotal := b.inputTotal + value
+
+	weight := b.weightEstimate.Weight()
+	fee := b.feePerKW.FeeForWeight(int64(weight))
+
+	// Calculate the output value if the current input would be
+	// added to the set.
+	newOutputValue := newInputTotal - fee
+
+	// If adding this input makes the total output value of the set
+	// decrease, this is a negative yield input. We don't add the input to
+	// the set and return the outcome.
+	if newOutputValue <= b.outputValue {
+		return false
+	}
+
+	// Update running values.
+	b.inputTotal = newInputTotal
+	b.outputValue = newOutputValue
+	b.inputs = append(b.inputs, input)
+
+	return true
+}
+
+// addPositiveYieldInputs adds sweepableInputs that have a positive yield to the
+// input set. This function assumes that the list of inputs is sorted descending
+// by yield.
+//
+// TODO(roasbeef): Consider including some negative yield inputs too to clean
+// up the utxo set even if it costs us some fees up front.  In the spirit of
+// minimizing any negative externalities we cause for the Bitcoin system as a
+// whole.
+func (b *txInputSet) addPositiveYieldInputs(sweepableInputs []txInput,
+	maxInputs int) {
+
+	for idx, input := range sweepableInputs {
+		// Try to add the input to the transaction. If that doesn't
+		// succeed because it wouldn't increase the output value,
+		// return. Assuming inputs are sorted by yield, any further
+		// inputs wouldn't increase the output value either.
+		if !b.add(input) {
+			return
+		}
+
+		// Stop if max inputs is reached.
+		if idx == maxInputs-1 {
```

Comment:
Shouldn't `maxInputs` be enforced to `b.inputs` if `addPositiveYieldInputs` is called multiple times to the same set?

Reply:
That doesn't happen at the moment, but changed this just in case.